apps/src/blockly/addons/plusMinusBlocks/advancedProcedures.js

```diff
@@ -777,34 +798,93 @@ GoogleBlockly.Extensions.register(
   'argument_report_get_var_models',
   function () {
     this.getVarModels = () => {
-      const varModels = [];
+      if (this.model) {
+        return [this.model];
+      }
+      if (this.isShadow()) {
+        return [
+          this.getRootBlock().argData_.find(
+            argDatum =>
+              argDatum.argId ===
+              this.outputConnection.targetConnection.getParentInput().name
+          ).model,
+        ];
+      }
+      // If we don't have a model yet, fall back on finding one based on the name.
       const foundModel = this.workspace.getVariable(this.getFieldValue('VAR'));
       if (foundModel) {
-        varModels.push(foundModel);
+        return [foundModel];
       }
-      return varModels;
+      console.error(`No model found for argument_reporter block`);
+      return [];
     };
   }
 );
 
-function deleteUnusedVariables(workspace) {
-  // Get all declared variables
-  const allVariables = workspace.getAllVariables();
-
-  // Get all used variables
-  const usedVariables = Blockly.Variables.allUsedVarModels(workspace);
-
-  // Convert used variables to a set for easy lookup
-  const usedVariableIds = new Set(
-    usedVariables.map(variable => variable.getId())
-  );
-
-  // Find unused variables
-  const unusedVariables = allVariables.filter(
-    variable => !usedVariableIds.has(variable.getId())
-  );
-  unusedVariables.forEach(varToDelete => {
-    console.log(`Deleting unused variable ${varToDelete.name}`);
-    workspace.deleteVariableById(varToDelete.getId());
+export function filterFunctionArgVariables(workspace, flyoutContents) {
+  return flyoutContents.filter(toolboxItem => {
+    const nonParamVarIds = getNonFunctionVariableIds(workspace);
+    // If we don't have any variables to show, just show the button to create one.
+    if (
+      !nonParamVarIds.length &&
+      toolboxItem.tagName.toLowerCase() !== 'button'
+    ) {
+      return false;
+    }
+
+    // For non-getter variables, select the first valid variable in the dropdown.
+    if (
+      toolboxItem.tagName.toLowerCase() === 'block' &&
+      toolboxItem.getAttribute('type') !== BLOCK_TYPES.variableGet
+    ) {
+      toolboxItem.firstElementChild?.setAttribute('id', nonParamVarIds[0]);
+    }
+    // Preserve all blocks except getters with invalid variables.
+    return (
+      toolboxItem.getAttribute('type') !== BLOCK_TYPES.variableGet ||
+      nonParamVarIds.includes(toolboxItem.firstElementChild.getAttribute('id'))
+    );
   });
 }
+
+// Finds or creates a variable model for a given name.
+function getModelForNewValue(block, varName) {
+  const workspace = block.workspace;
+  // Create new vars instead of renaming the old ones, so users can't
+  // accidentally rename/coalesce vars.
```

Comment:
I'm slightly confused by this function's comments. It first says we create new vars instead of renaming old ones, but then later we rename the variable. Am I missing some distinction between the cases?

Reply:
Yeah, it's confusing, but I think I can break it down. In Blockly, all variables are global. So if two functions both have a parameter `x`, then they're using the same variable. If a user types `y` to "rename" one parameter, we don't actually want to rename the existing `x` variable, because that would rename it everywhere. So instead, we make a new variable, or potentially we find an existing `y` variable to use if it already exists.

The renaming comment below has to do with case-insensitivity. `getVariable('y', '')` will actually find and return a variable called `y` or `Y`, if one exists. In this case, we _do_ rename so that what the user actually types becomes the new variable name.